### PR TITLE
Video uploads with google drive API

### PR DESCRIPTION
**Pull request by JoshGuempel**
Backend sends video uploads to google drive with a service account.

### CODE REVIEW

**Review by ClayMav — CHANGES_REQUESTED** [review]
Permissions are a big problem here, users on google drive need to be able to see the files. Also dont ask for a mimeType, just see what mimeType is provided and ensure that it is a video file. Test with a shared drive and make sure it works with one of those too. Squash and merge will be the move once this is done and approved

**Comment by JoshGuempel** [comment]
Updated file uploads to work with shared drives, and also removed mimeType parameter from the video upload mutation.

File: api/file_manager/file_manager.ts [hunk 1]
```diff
@@ -12,7 +12,10 @@ export async function createFile(
     fileStream: fs.ReadStream
 ): Promise<string> {
     const JWTAuth: JWT = await getJWTAuth()
-    const drive = google.drive({ version: 'v3', auth: JWTAuth })
+    const drive: drive_v3.Drive = google.drive({
+        version: 'v3', 
+        auth: JWTAuth
+    })
```
Comment (ClayMav) [inline api/file_manager/file_manager.ts]: I'm not sure you need to define the drive every time you do an action, why not create it globally and use it when you need it

File: api/file_manager/file_manager.ts [hunk 2]
```diff
@@ -39,23 +44,27 @@ export async function createFile(
 
 export async function createFolder(folderName: string): Promise<string> {
     const JWTAuth: JWT = await getJWTAuth()
-    const drive = google.drive({ version: 'v3', auth: JWTAuth }) 
+    const drive: drive_v3.Drive = google.drive({
+        version: 'v3', 
+        auth: JWTAuth
+    })
```
Comment (ClayMav) [inline api/file_manager/file_manager.ts]: as above

File: api/file_manager/file_manager.ts [hunk 3]
```diff
@@ -64,9 +73,15 @@ export async function createFolder(folderName: string): Promise<string> {
 //this is for debugging purposes.
 export async function listFiles(): Promise<void> {
     const JWTAuth: JWT = await getJWTAuth()
-    const drive = google.drive({ version: 'v3', auth: JWTAuth })
+    const drive: drive_v3.Drive = google.drive({
+        version: 'v3', 
+        auth: JWTAuth
+    })
```
Comment (ClayMav) [inline api/file_manager/file_manager.ts]: as above

File: api/resolvers/video-resolver.ts [hunk 4]
```diff
@@ -13,9 +13,7 @@ export class VideoResolver {
     @Mutation(() => Video)
     protected async uploadVideo(
         @Arg('video', () => GraphQLUpload)
-        { createReadStream, filename }: Upload,
-        @Arg('mimeType', () => String)
-        mimeType: string,
+        { createReadStream, filename, mimetype }: Upload,
```
Comment (ClayMav) [inline api/resolvers/video-resolver.ts:18]: I would do a check to make sure that the mimetype is a video, check for the video files output from ios and android specifically as well as the common mp4, mov, etc

**Review by ClayMav — CHANGES_REQUESTED** [review]

File: api/file_manager/file_manager.ts [hunk 5]
```diff
@@ -1,21 +1,23 @@
...
-    const drive: drive_v3.Drive = google.drive({
-        version: 'v3', 
-        auth: JWTAuth
-    })
+    //Make sure video is a video
+    if(!videoMimeTypes.includes(mimeType)) {
+        throw new Error(
+            'File mime type not recognized as a video.'
+        )
+    }
```
Comment (ClayMav) [inline api/file_manager/file_manager.ts]: definitely move all this logic to the resolver for the video

File: api/resolvers/video-resolver.ts [hunk 6]
```diff
@@ -19,18 +19,22 @@ export class VideoResolver {
...
+            var videoId: string = await createFile(filename, mimetype, flightDriveId, createReadStream())
+            const video: Video = this.videoRepo.create({
+                driveId: videoId
+            })
+            video.flight = flight
+            
+            return video.save()
+        } catch (err) {
+            console.log(err)
+        }
```
Comment (ClayMav) [inline api/resolvers/video-resolver.ts]: Dont catch it, let it blow up

**Review by ClayMav — CHANGES_REQUESTED** [review]

File: api/resolvers/video-resolver.ts [hunk 7]
```diff
@@ -17,29 +19,48 @@ export class VideoResolver {
...
+
+    @Mutation(() => Boolean)
+    protected async deleteVideoByID(
+        @Arg('id', () => Int)
+        id: number
+    ): Promise<boolean> {
+        let video: Video = await this.videoRepo.findOneOrFail({
+            id: id,
+        })
+        await deleteFolder(video.driveId)
```
Comment (ClayMav) [inline api/resolvers/video-resolver.ts]: its a video not a folder, maybe rename the function

**Review by ClayMav — CHANGES_REQUESTED** [review]

**Review by ClayMav — APPROVED** [review]

**Pull request merged**